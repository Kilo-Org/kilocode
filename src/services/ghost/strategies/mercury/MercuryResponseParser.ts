/**
 * Handles parsing and cleaning Mercury's responses to extract the updated code.
 * Mercury responses can come in various formats (markdown, with/without markers).
 */
export class MercuryResponseParser {
	private static readonly MERCURY_MARKERS = {
		OPEN: "<|code_to_edit|>",
		CLOSE: "<|/code_to_edit|>",
	}

	/**
	 * Extract clean code content from Mercury's response.
	 * Handles multiple fallback strategies for robust parsing.
	 */
	public extractCleanCode(response: string): string {
		// Step 1: Extract from markdown code blocks first
		let extractedCode = this.extractFromCodeBlocks(response)

		// Step 2: Handle Mercury markers (fallback in case they leak through)
		extractedCode = this.stripMercuryMarkers(extractedCode)

		return extractedCode.trim()
	}

	/**
	 * Extract content from markdown code blocks (```language code ```)
	 */
	private extractFromCodeBlocks(message: string): string {
		// Look for code blocks with optional language specifiers
		const codeBlockRegex = /```(?:\w+)?\s*\n?([\s\S]*?)\n?```/g
		const matches = codeBlockRegex.exec(message)

		if (matches && matches[1]) {
			return matches[1].trim()
		}

		// Fallback: try simpler patterns
		const startMarker = "```"
		const startIndex = message.indexOf(startMarker)
		const endIndex = message.lastIndexOf(startMarker)

		if (startIndex !== -1 && endIndex !== -1 && startIndex !== endIndex) {
			// Find the end of the first line after the opening ```
			let contentStart = message.indexOf("\n", startIndex)
			if (contentStart === -1) {
				contentStart = startIndex + 3
			} else {
				contentStart += 1
			}

			const content = message.slice(contentStart, endIndex).trim()
			return content
		}

		// No code blocks found, return the original message
		return message.trim()
	}

	/**
	 * Strip Mercury markers from content.
	 * These shouldn't be in the response, but add fallback in case they leak through.
	 * Also handles line number stripping while preserving indentation.
	 */
	private stripMercuryMarkers(content: string): string {
		const { OPEN, CLOSE } = MercuryResponseParser.MERCURY_MARKERS

		// Remove opening marker
		let cleaned = content.replace(new RegExp(OPEN, "g"), "")

		// Remove closing marker
		cleaned = cleaned.replace(new RegExp(CLOSE.replace(/[|]/g, "\\|"), "g"), "")

		// CRITICAL FIX: Remove line numbers while preserving indentation
		// Pattern: "123 | content" -> "content" (keeping all spaces after |)
		cleaned = this.stripLineNumbersPreservingIndentation(cleaned)

		return cleaned
	}

	/**
	 * Strip line numbers from Mercury response while preserving indentation.
	 * Fixes the critical whitespace loss bug.
	 */
	private stripLineNumbersPreservingIndentation(content: string): string {
		// Regex to match: [digits] [space] [|] [content including spaces]
		// Captures everything after "| " to preserve indentation
		return content.replace(/^\d+\s*\|\s?/gm, "")
	}

	/**
	 * Check if response appears to be empty or just whitespace
	 */
	public isEmpty(response: string): boolean {
		return !this.extractCleanCode(response).trim()
	}

	/**
	 * Get debug info about the parsing process
	 */
	public getParsingDebugInfo(response: string): {
		originalLength: number
		hasCodeBlocks: boolean
		hasMercuryMarkers: boolean
		extractedLength: number
	} {
		const { OPEN, CLOSE } = MercuryResponseParser.MERCURY_MARKERS

		return {
			originalLength: response.length,
			hasCodeBlocks: response.includes("```"),
			hasMercuryMarkers: response.includes(OPEN) || response.includes(CLOSE),
			extractedLength: this.extractCleanCode(response).length,
		}
	}
}
