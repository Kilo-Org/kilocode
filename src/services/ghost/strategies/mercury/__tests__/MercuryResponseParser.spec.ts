import { describe, it, expect } from "vitest"
import { MercuryResponseParser } from "../MercuryResponseParser"

describe("MercuryResponseParser", () => {
	let parser: MercuryResponseParser

	beforeEach(() => {
		parser = new MercuryResponseParser()
	})

	describe("extractCleanCode", () => {
		it("should extract code from markdown code blocks with language", () => {
			const response = `Here's the updated code:

\`\`\`javascript
function hello() {
		  console.log("Hello World");
}
\`\`\``

			const extracted = parser.extractCleanCode(response)
			expect(extracted).toMatchInlineSnapshot(`
				"function hello() {
						  console.log("Hello World");
				}"
			`)
		})

		it("should extract code from code blocks without language specifier", () => {
			const response = `\`\`\`
const x = 42;
const y = 13;
\`\`\``

			const extracted = parser.extractCleanCode(response)
			expect(extracted).toMatchInlineSnapshot(`
				"const x = 42;
				const y = 13;"
			`)
		})

		it("should handle Mercury markers if they leak through", () => {
			const response = `<|code_to_edit|>
function test() {
		  return true;
}
<|/code_to_edit|>`

			const extracted = parser.extractCleanCode(response)
			expect(extracted).toMatchInlineSnapshot(`
				"||
				function test() {
						  return true;
				}
				|/|"
			`)
		})

		it("should handle combination of markdown and mercury markers", () => {
			const response = `\`\`\`javascript
<|code_to_edit|>
const value = "test";
<|/code_to_edit|>
\`\`\``

			const extracted = parser.extractCleanCode(response)
			expect(extracted).toMatchInlineSnapshot(`
				"||
				const value = "test";
				|/|"
			`)
		})

		it("should return original content if no markers found", () => {
			const response = `function plain() { return 42; }`

			const extracted = parser.extractCleanCode(response)
			expect(extracted).toMatchInlineSnapshot(`"function plain() { return 42; }"`)
		})

		it("should preserve whitespace and indentation perfectly", () => {
			const response = `\`\`\`typescript
class User {
		  constructor(name, age) {
		      this.name = name;
		      this.age = age;
		  }

		  getDisplayName() {
		      return \`Name: \${this.name}, Age: \${this.age}\`;
		  }
}
\`\`\``

			const extracted = parser.extractCleanCode(response)
			expect(extracted).toMatchInlineSnapshot(`
				"class User {
						  constructor(name, age) {
						      this.name = name;
						      this.age = age;
						  }

						  getDisplayName() {
						      return \`Name: \${this.name}, Age: \${this.age}\`;
						  }
				}"
			`)
		})
	})

	describe("isEmpty and debug info", () => {
		it("should detect empty vs non-empty responses", () => {
			const testCases = [
				{ input: "", desc: "empty string" },
				{ input: "   \n\t  ", desc: "whitespace only" },
				{ input: "```\n\n```", desc: "empty code block" },
				{ input: "const x = 1;", desc: "plain code" },
				{ input: "```\nconst x = 1;\n```", desc: "code in block" },
			]

			const results = testCases.map(({ input, desc }) => ({
				desc,
				isEmpty: parser.isEmpty(input),
				debugInfo: parser.getParsingDebugInfo(input),
			}))

			expect(results).toMatchSnapshot()
		})
	})
})
