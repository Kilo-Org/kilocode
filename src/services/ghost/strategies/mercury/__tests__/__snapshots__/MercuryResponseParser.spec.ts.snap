// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MercuryResponseParser > isEmpty and debug info > should detect empty vs non-empty responses 1`] = `
[
  {
    "debugInfo": {
      "extractedLength": 0,
      "hasCodeBlocks": false,
      "hasMercuryMarkers": false,
      "originalLength": 0,
    },
    "desc": "empty string",
    "isEmpty": true,
  },
  {
    "debugInfo": {
      "extractedLength": 0,
      "hasCodeBlocks": false,
      "hasMercuryMarkers": false,
      "originalLength": 7,
    },
    "desc": "whitespace only",
    "isEmpty": true,
  },
  {
    "debugInfo": {
      "extractedLength": 0,
      "hasCodeBlocks": true,
      "hasMercuryMarkers": false,
      "originalLength": 8,
    },
    "desc": "empty code block",
    "isEmpty": true,
  },
  {
    "debugInfo": {
      "extractedLength": 12,
      "hasCodeBlocks": false,
      "hasMercuryMarkers": false,
      "originalLength": 12,
    },
    "desc": "plain code",
    "isEmpty": false,
  },
  {
    "debugInfo": {
      "extractedLength": 12,
      "hasCodeBlocks": true,
      "hasMercuryMarkers": false,
      "originalLength": 20,
    },
    "desc": "code in block",
    "isEmpty": false,
  },
]
`;
